src/sql/src/optimizer_metrics.rs

```diff
@@ -40,10 +40,23 @@ impl OptimizerMetrics {
         }
     }
 
-    pub fn observe_e2e_optimization_time(&self, object_type: &str, duration: Duration) {
+    pub fn observe_e2e_optimization_time<F: Fn() -> String>(
+        &self,
+        object_type: &str,
+        duration: Duration,
+        show_query: F,
+    ) {
         self.e2e_optimization_time_seconds
             .with_label_values(&[object_type])
-            .observe(duration.as_secs_f64())
+            .observe(duration.as_secs_f64());
+        if duration > Duration::from_millis(500) {
+            tracing::warn!(
+                object_type = object_type,
+                query = show_query(),
```

Comment:
Just to double check, we don't include `show_query()` anywhere in our Prometheus metrics? If we did that would blow up cardinality which AFAIU Prometheus does not handle well

Reply:
No, not in Prometheus. But in any case, I removed this part, since we've already attached the redacted query appropriately.